### PR TITLE
Reimplement hazelcast hibernate caching

### PR DESCRIPTION
**Pull request by cszmajda**
# tl;dr

The previous implementation of Hibernate L2 caching had incorrect transactional behavior and would cache stale values and/or fail with "IllegalMonitorStateException: current thread not owner of the lock" under heavy load.

This PR re-implements the Hibernate region cache to implement Hibernate "soft locking" correctly. The implementation is loosely based on Ehcache (Apache 2 licensed), which should be considered the "reference" implementation of Hibernate's L2 caching SPI. Calls to `IMap.lock` have been replaced wherever possible with CAS or entry processors, resulting in more efficient and better transactional semantics, even in edge cases such as server crashing and network partitions.

We've been running this implementation in production for months, and have subjected it to extremely heavy stress testing in our cluster lab (up to 400 threads spread over 4 high-spec machines subjecting a cluster of 1-4 nodes to continuous hibernate-intensive load). Problems we had in the previous implementation due to stale data and "IllegalMonitorStateException: current thread not owner of the lock" errors have been completely eliminated. 

# Soft locking

The main difference in this implementation over the old one is that this one correctly implements how Hibernate's soft locks are suppose to behave. There is some documentation on this behavior in the [JavaDoc for ExpiryMarker](https://github.com/hazelcast/hazelcast/pull/4110/files#diff-75a485177b9e1e7dcf639fda2844cd5fR27). Soft-locking requires the following characteristics:

* Multiple transactions (i.e. threads) can soft lock an entry concurrently
* An entry isn't unlocked until all transactions have completed and unlocked the entry
* Soft locks have an expiry/timeout timestamp to handle the case where an transaction fails to unlock after locking (e.g server crash, network partition)
* The entry is considered null (i.e. a cache miss) for the duration of it being locked

The previous implementation did not meet these requirements:

* It attempted to lock an entry exclusively for a set timeout, if it failed to lock the entry it did not attempt to unlock it
* It did not report the cache entry as empty when it is soft locked, causing stale cache values to be used by hibernate when determining what it should do
* Entries were not unlocked in the case of server crashing. 
* Locks were silently released in the event of a network partition that later heals.  Notice that `ClusterServiceImpl.merge` resets all managedServices including every lockStore. This means that all locks held are effectively released, invariably leading to `java.lang.IllegalMonitorStateException: Current thread is not owner of the lock! -> <not-locked>`

### CODE REVIEW

**Comment by bilalyasar** [comment]
Can one of the admins verify this patch?

**Comment by pveentjer** [comment]
I'm going to close this one due to inactivity. Feel free to re-open it.

**Comment by cszmajda** [comment]
Hi @pveentjer, The code changes in this PR are still very much essential for our use of the Hibernate integration in Hazelcast.  I've noticed other Hazelcast users are still running into the same errors (`IllegalMonitorStateException` and the like) that we solved quite some time ago in this PR.  What would it take to move this forward?

**Comment by enesakar** [comment]
Hi @pveentjer @cszmajda ;
I talked with integration team (owner of this module internally), they have plans to merge this PR.
So I am reopening this.

**Comment by cszmajda** [comment]
I fixed a couple of small conflicts in this PR, it should still be good to go.

**Comment by noctarius** [comment]
Hey @cszmajda, thanks again for still updating this one. Getting such essential changes into the system is not as easy as we all sometimes would like. Integration team is still looking at this PR so stay tuned :)

Still an amazing effort!

**Comment by bwzhang2011** [comment]
@kobalski, could it merged into branch3.5, we're looking for better integration with hibernate

**Comment by cszmajda** [comment]
I just fixed the checkstyle issues and, while I was there, merged in the latest `master` from upstream.  This passes all tests for me locally.

File: hazelcast-hibernate/hazelcast-hibernate4/src/main/resources/META-INF/services/com.hazelcast.NodeInitializer [hunk 1]
```diff
@@ -0,0 +1 @@
+com.hazelcast.instance.ExtendedNodeInitializer
```
Comment (kobalski) [inline hazelcast-hibernate/hazelcast-hibernate4/src/main/resources/META-INF/services/com.hazelcast.NodeInitializer]: @cszmajda I think we should remove this file, because `NodeInitializer` is hazelcast-2x class https://github.com/hazelcast/hazelcast/blob/maintenance-2.x/hazelcast/src/main/java/com/hazelcast/impl/base/NodeInitializer.java.  Also, why did you need `ExtendedNodeInitializer` ?

Comment (bturner) [inline hazelcast-hibernate/hazelcast-hibernate4/src/main/resources/META-INF/services/com.hazelcast.NodeInitializer]: Sorry, that's a mistake. It shouldn't be included in this pull request.

We use custom `NodeExtension`, in the new parlance, to remove the `WanReplicationService`, which otherwise starts 16 threads and never uses them. Since WAN replication is an enterprise feature, it would be nice if there was a simple way to turn it off. Or, even better, if it was just off automatically when you don't have a license for it. Since it's not, we override `createService` to return a no-op `WanReplicationService` implementation.

Comment (mdogan) [inline hazelcast-hibernate/hazelcast-hibernate4/src/main/resources/META-INF/services/com.hazelcast.NodeInitializer]: @bturner,

Since v3.4 (https://github.com/hazelcast/hazelcast/commit/7f826862ecc8e8308eeaf928dd5499595f2a4cd0) `WanReplicationService` creates/starts its threads on-demand. By default, when not used/touched, wan threads won't be created.  

Btw, `WanReplicationService` is not a completely enterprise feature, current open-source versions will remain as it is. Just new/enhanced version is enterprise.

**Comment by alex-pozdnyakov** [comment]
@cszmajda, in your implementation method [nextMarkerId()](https://github.com/atlassian/hazelcast/blob/reimplement-hazelcast-hibernate-caching/hazelcast-hibernate/hazelcast-hibernate4/src/main/java/com/hazelcast/hibernate/distributed/IMapRegionCache.java#L158) from [IMapRegionCache](https://github.com/atlassian/hazelcast/blob/reimplement-hazelcast-hibernate-caching/hazelcast-hibernate/hazelcast-hibernate4/src/main/java/com/hazelcast/hibernate/distributed/IMapRegionCache.java) calls [hazelcastInstance.getCluster().getLocalMember()](https://github.com/atlassian/hazelcast/blob/reimplement-hazelcast-hibernate-caching/hazelcast-hibernate/hazelcast-hibernate4/src/main/java/com/hazelcast/hibernate/distributed/IMapRegionCache.java#L159), but this method in [ClientClusterProxy](https://github.com/atlassian/hazelcast/blob/reimplement-hazelcast-hibernate-caching/hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientClusterProxy.java#L57) throws UnsupportedOperationException and can't be used for native client.

**Comment by cszmajda** [comment]
> nextMarkerId() from IMapRegionCache calls hazelcastInstance.getCluster().getLocalMember()

@alex-pozdnyakov Thanks for pointing that out, I've replaced this with getLocalEndpoint().

**Comment by cszmajda** [comment]
I've also just pushed the changes that bring LOCAL mode Hibernate caching up to the same level as DISTRIBUTED mode.  The previous implementation did not correctly invalidate Collections of related entities when these were updated.  LOCAL mode trades some consistency but can be a good option for users with very heavy Hibernate (especially transactional) load.  See 8f53ade above for details.

**Comment by alex-pozdnyakov** [comment]
@cszmajda, i have next call stack trace:
```
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: No DataSerializerFactory registered for namespace: -2000
	at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:98)
	at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:39)
	at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:41)
	at com.hazelcast.nio.serialization.SerializationServiceImpl.readObject(SerializationServiceImpl.java:325)
	at com.hazelcast.nio.serialization.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:489)
	at com.hazelcast.map.impl.client.MapExecuteOnKeyRequest.read(MapExecuteOnKeyRequest.java:93)
	at com.hazelcast.client.impl.client.ClientRequest.readPortable(ClientRequest.java:97)
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:84)
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:73)
	at com.hazelcast.nio.serialization.PortableSerializer.read(PortableSerializer.java:29)
	at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:41)
	at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:276)
	at com.hazelcast.client.impl.ClientEngineImpl$ClientPacketProcessor.loadRequest(ClientEngineImpl.java:391)
	at com.hazelcast.client.impl.ClientEngineImpl$ClientPacketProcessor.run(ClientEngineImpl.java:366)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:100)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPartitionSpecificRunnable(OperationThread.java:130)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:120)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:101)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:76)
	at ------ End remote and begin local stack-trace ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveResponse(ClientInvocationFuture.java:147)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.get(ClientInvocationFuture.java:114)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.get(ClientInvocationFuture.java:89)
	at com.hazelcast.client.spi.ClientProxy.invoke(ClientProxy.java:130)
	at com.hazelcast.client.proxy.ClientMapProxy.executeOnKey(ClientMapProxy.java:939)
	at com.hazelcast.hibernate.distributed.IMapRegionCache.unlock(IMapRegionCache.java:128)
	at com.hazelcast.hibernate.access.ReadWriteAccessDelegate.unlockItem(ReadWriteAccessDelegate.java:57)
	at com.hazelcast.hibernate.region.EntityRegionAccessStrategyAdapter.unlockItem(EntityRegionAccessStrategyAdapter.java:94)
	at org.hibernate.action.internal.EntityUpdateAction.doAfterTransactionCompletion(EntityUpdateAction.java:342)
	at org.hibernate.engine.spi.ActionQueue$AfterTransactionCompletionProcessQueue.afterTransactionCompletion(ActionQueue.java:751)
	... 88 more
```
No hazelcast.serialization.ds.hibernate option is set.

**Comment by cszmajda** [comment]
@alex-pozdnyakov I'm not sure what's causing this.  The hazelcast.serialization.ds.hibernate factory is declared here
https://github.com/hazelcast/hazelcast/pull/4110/files#diff-8c3167c53fc31aa3ab50f0efe7dce04cR42 and we are using this all the time in production, though we use full Hazelcast instances, not clients.  Can you share your test code for the above exception?

**Comment by kobalski** [comment]
verify

**Comment by kobalski** [comment]
verify

**Comment by mesutcelik** [comment]
verify

**Comment by kobalski** [comment]
@alex-pozdnyakov I also tested with native client mode and did not see any exceptions.  Please make sure that both remote cluster and client used by L2 cache has the same hazelcast version. Other than that, I think that this PR is mergable to me. I must also mention that, we will have verification process for the new design after merge. Meanwhile, if you see problems with the new design according to your use case scenario, feel free to report.

**Comment by mesutcelik** [comment]
HI @cszmajda ,

Thanks for great contribution. It seems this PR is mergeable but i see some commits from @jhinch . 
Can we get contributor agreement from him as well?

https://hazelcast.atlassian.net/wiki/display/COM/Hazelcast+Contributor+Agreement

**Comment by cszmajda** [comment]
@mesutcelik I've passed that on to @jhinch who deserves the real credit for this contribution.  We've also sent through the executed contributor agreement via email as per your instructions.

**Comment by alex-pozdnyakov** [comment]
@cszmajda @kobalski [test app](https://github.com/alex-pozdnyakov/hazelcast-hibernate-native-client-test) to reproduce my case. This works fine with Ehcache, but crashes with Hazelcase native client configured. I'm use custom builded Hazelcast with [this revision](https://github.com/hazelcast/hazelcast/commit/4270dc61e315b82c1b6db4f65851c03eba595a83) (it's 3.5 version) and patch from this pull request, so Hazelcast version in pom.xml is 3.5. To run application, run **mvn exec:java -Dexec.mainClass="test.App"** command.

**Comment by jhinch** [comment]
Good to see this finally merged. I understand it was a big change and we appreciate you working with us on it. @alex-pozdnyakov we only used the implementation server mode. However it _should_ work in client mode as well. From the exception you reported, it seems like the Hazelcast server doesn't have hazelcast-hibernate on its classpath. This is very much a requirement of this change due to the heavy use of EntityProcessors for various operations. Can you confirm whether or not the Hazelcast server has hazelcast-hibernate on the classpath or not. From the repository you linked, it wasn't obvious how the server was configured.

**Comment by alex-pozdnyakov** [comment]
@jhinch Looks like you are right. But when i put hazelcast-hibernate4 into classpath and run my sample app, i get
```Caused by: java.lang.ClassNotFoundException: org.hibernate.cache.spi.access.SoftLock
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.hazelcast.hibernate.serialization.HibernateDataSerializerHook$Factory.create(HibernateDataSerializerHook.java:94)
	at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:101)
	at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:39)
	at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:41)
	at com.hazelcast.nio.serialization.SerializationServiceImpl.readObject(SerializationServiceImpl.java:325)
	at com.hazelcast.nio.serialization.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:489) 
     .....```

Ok, i put hibernate-core into classpath too and get

```Caused by: java.lang.ClassNotFoundException: test.entity.UserValuePk
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:125)
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:114)
	at com.hazelcast.nio.IOUtil$1.resolveClass(IOUtil.java:113)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1613)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1518)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1774)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
	at com.hazelcast.nio.serialization.DefaultSerializers$ObjectSerializer.read(DefaultSerializers.java:196)
	... 27 more```

test.entity.UserValuePk is my entity class from test application. Why standalone Hazelcast server need for additional jars and **my application's classes** in **server** classpath?

**Comment by jhinch** [comment]
Fundamentally the algorithm for the distributed version of the Hibernate cache relies on EntryProcessors to work in a way that scales. It _could_ be implemented without it using either locking compare and set in a loop, but in developing this implementation for use in Atlassian Stash, we found that locking operations failed too often under high load and a compare and set in a loop would result in the implementation being far too chatty with the server.

I think that the local cache implementation (with cache invalidation via Hazelcast) as a good option for the client mode. Another potential option would be to lazily deserialize the cached value in the entry processors and Expirable. This may remove the need to deserialize the value on the server.

**Comment by alex-pozdnyakov** [comment]
So, as i understand its correctly, no way to configure HazelcastCacheRegionFactory without third party jars and client's entites jars exists?

**Comment by jhinch** [comment]
With a standalone Hazelcast server and using the Hazelcast client, with this change merged as the code currently is, you're only option is to use LOCAL mode for the region cache OR add the third party jar to the servers classpath.

It works perfectly fine if you are embedding the Hazelcast server in your application. This is what Atlassian Stash is doing. When I re-implemented it, I didn't even realize that consuming a Hazelcast client based implementation of the region cache was possible. As a result I didn't account for it in my implementation.

However as I mentioned in my last comment I think there is a reasonable path forward which would restore support for a client based set up. The entry processors just need to be able to check is the underlying value is null, it doesn't actually need to be able to deserialize the value. As a result, if the value was lazily deserialized, it could be avoided entirely within the entry processors and only deserialized when it is returned to the client.

**Comment by kobalski** [comment]
@alex-pozdnyakov Yes, you are right with the use of EntryProcessor in distributed map requires hibernate core jar to be present in the standalone server's classpath. But you should not need your entity classes in the classpath of the server. I think this [comment](https://github.com/hazelcast/hazelcast/pull/4110/#issuecomment-63748385) clearly explains the situation.  Here I have created a test project (https://github.com/kobalski/hibernate-test-cluster), which starts two node stand alone server which has hazelcast and hibernate-core jars and after that I run your test app. I did not see entity related exceptions. Just like you said, if I remove hibernate jars. I get class not found exception for ` org.hibernate.cache.spi.access.SoftLock` class and if I remove hazelcast-hibernate4 jars , I get serialisation exception `No DataSerializerFactory registered for namespace: -2000` but I could not reproduce `java.lang.ClassNotFoundException: test.entity.UserValuePk` case.  If entity classes are needed on the classpath of the server, we need to fix that. Also since this PR is merged, you can use `3.6-SNAPSHOT` for hazelcast version in your test app, I think maven repos should be updated with current code by now. So can you try to reproduce the issue  with [this](https://github.com/kobalski/hibernate-test-cluster) server project?

**Comment by alex-pozdnyakov** [comment]
@kobalski It's strange, but it works on 3.6-SNAPSHOT. Previously, i use server [from here](http://download.hazelcast.com/download.jsp?version=hazelcast-3.5&type=tar&p=108073472) with patched jars in lib folder. Any diffrerences between run configurations?

**Comment by kobalski** [comment]
@alex-pozdnyakov I do not know the situation maybe sth wrong with the patch, right now 3.6-SNAPSHOT has latest code, if it is working we have no problem. Also, Hibernate 2nd Level Cache uses [CacheEntry](http://grepcode.com/file/repo1.maven.org/maven2/org.hibernate/hibernate-core/3.6.0.Final/org/hibernate/cache/entry/CacheEntry.java?av=f) class to store entity values inside the cache provider. You should not need to have entity classes on the server side. But current design has one flaw, if you use client mode with distributed mode(HazelcastCacheRegionFactory), you need to have hibernate core jars on the class path of the server. Imagine a scenario where a big hazelcast cluster(server) is used by many hazelcast clients. One team is using hibernate 3 and another team is hibernate 4, then in this scenario you need to have both hibernate3 and hibernate4 on the classpath if both projects wants to use hazelcast cluster for 2nd Level Cache and probably it will cause classloader problems. So for now, if you have this kind of scenario, I recommend as @jhinch said, use LOCAL mode(HazelcastLocalCacheRegionFactory) instead of distributed node.
